### DISCUSSION TITLE
TypeGraphQL Example Integration

### DISCUSSION THREAD

**Discussion by LaCocoRoco** [vercel/next.js#34398] [Ideas]
### What example does this report relate to?

graphql

### What version of Next.js are you using?

12.0.10

### What version of Node.js are you using?
16.13.0

### What browser are you using?

Chrome

### What operating system are you using?

Linux

### How are you deploying your application?

next start

### Describe the Bug

I could not solve the problem to use the decorator class like **InputType()** or **ArgsType()** in the **Resolver()** class decorators like **recipes(Args() { first, last }: RecipesArgs)** without getting the following error:

- **Unable to infer GraphQL type from TypeScript reflection system. You need to provide explicit type for parameter...**

This particular error i could fix on other decorators by defining their specific type from **Field()** to **Field(type => String)**.
It is probably possible to fix the error on the decorator class by defining their type, but ath the end i did not find a workable solution. I am also new to decorators. At some point i also suspect some problems under the hood in the nextjs package witch probably change the behaviour of the type reflection.

### Expected Behavior

The request should not fail with the "TypeScript reflection error" while requesting data via api.

### To Reproduce

I prepared a repository to reproduce the problem: https://github.com/LaCocoRoco/api-routes-apollo-server-type-graphql
With **npm run dev** a graphql request can be made at: http://localhost:3000/api/graphql
With **npm run generate** it is possible to generate a **schema.gql**
In the **tsconfig.json** i changed the module to **commonjs**
The configuration did not change the behaviour of the problem itself

**Comment by pabl-o-ce** [comment]
Hi @LaCocoRoco 
Are you trying to make Graphql API into nextjs with success? [I was also trying to make this but then I ended spliting the project]

**Comment by LaCocoRoco** [comment]
NextJS already has a working example for GraphQL
https://github.com/vercel/next.js/tree/canary/examples/api-routes-apollo-server-and-client

If you want to use the TypeGraphQL package, then you could use this as reference.
https://github.com/LaCocoRoco/api-routes-apollo-server-type-graphql

Regarding this issue:
I fixed all the errors and problems, but will probably remove the example from my repository.
To integrate the type-graphql package, i changed the following points.

```typescript
const apolloServerInitialized = (async () => {
  // build schema before initializing apollo server
  const apolloServer = new ApolloServer({
    schema: await schema(),
    context: (context: Context): Context => context
  });

  await apolloServer.start();

  return apolloServer;
})();

const apolloServer = async (req: NextApiRequest, res: NextApiResponse) => {
  const apolloServer = await apolloServerInitialized;

  const apolloHandler = apolloServer.createHandler({
    path: req.url,
  });

  await apolloHandler(req, res);
}

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  await apolloServer(req, res);
}
```
The schema will be build before initializing the server.
With this i also fixed the error **MaxListenersExceededWarning: Possible EventEmitter memory leak detected**

It is also necessary to explicit define the type of the decorators.
In conclusion as example i had to change them from
```typescript
  @Mutation(() => Boolean)
  addRecipe(
    @Arg("recipeInputs",) recipeInputs: RecipeInputs
  ): boolean {
    return true;
  }

  @Query(() => [Recipe])
  getRecipeByRange(
    @Args() recipesArgs: RecipesArgs
  ): Recipe[] {
    return this.recipesCollection.slice(recipesArgs.first, recipesArgs.last);
  }
```
to
```typescript
  @Mutation(() => Boolean)
  addRecipe(
    @Arg("recipeInputs", () => RecipeInputs) recipeInputs: RecipeInputs
  ): boolean {
    return true;
  }

  @Query(() => [Recipe])
  getRecipeByRange(
    @Args(() => RecipesArgs) recipesArgs: RecipesArgs
  ): Recipe[] {
    return this.recipesCollection.slice(recipesArgs.first, recipesArgs.last);
  }
```
I added cors and ingerated jwt auth to the project based on the TypeGraphQL concept.

Probalby for future use an additional example for TypeGraphQL integration would be the way to go,
but this is a point to be handled by the nextjs devs.
I will close this issue the next few days.

**Comment by pabl-o-ce** [comment]
@LaCocoRoco thanks for the reference :) 👍
